plugins/out_s3/s3_multipart.c

```diff
@@ -636,14 +636,40 @@ int upload_part(struct flb_s3 *ctx, struct multipart_upload *m_upload,
     }
     uri = tmp;
 
+    struct flb_aws_header *headers = NULL;
+    int num_headers = 0;
+
+    char body_md5[25];
+    memset(body_md5, 0, sizeof(body_md5));
+    if (ctx->send_content_md5 == FLB_TRUE) {
+        if (s3_body_md5(body, body_size, body_md5, sizeof(body_md5)) != 0) {
+            flb_plg_error(ctx->ins, "Failed to MD5 body");
+            flb_sds_destroy(uri);
+            return -1;
+        }
+
+        num_headers = 1;
+        headers = flb_malloc(sizeof(struct flb_aws_header) * num_headers);
+        if (headers == NULL) {
+            flb_errno();
+            flb_sds_destroy(uri);
+            return -1;
+        }
+
+        headers[0].key = "Content-MD5";
+        headers[0].key_len = 11;
+        headers[0].val = body_md5;
+        headers[0].val_len = strlen(body_md5);
+    }
+
     s3_client = ctx->s3_client;
     if (s3_plugin_under_test() == FLB_TRUE) {
         c = mock_s3_call("TEST_UPLOAD_PART_ERROR", "UploadPart");
     }
     else {
         c = s3_client->client_vtable->request(s3_client, FLB_HTTP_PUT,
                                               uri, body, body_size,
-                                              NULL, 0);
+                                              headers, num_headers);
```

Comment:
Should the header be added in function `create_multipart_upload()`? https://github.com/fluent/fluent-bit/blob/master/plugins/out_s3/s3_multipart.c#L518

Reply:
Content-MD5 is required for PutObject and UploadPart. CreateMultipartUpload requests don't have a body, so the requirement doesn't apply.

Reply:
Thanks for the efforts! Just want to confirm that the code has been tested in two cases right? (PutObject & MultipartUpload)

Reply:
I'm still working on valgrind & testing multipart.

Reply:
Testing succeeded for UploadPart and PutObject, and valgrind is clean for both paths.